include/aws/io/retry_strategy.h

```diff
@@ -0,0 +1,175 @@
+    /** This is an error where the server explicitly told the client to back off, such as a 429 or 503 Http error. */
+    AWS_RETRY_ERROR_TYPE_THROTTLING,
+    /** This is a server error that isn't explicitly throttling but is considered by the client
+     * to be something that should be retried. */
+    AWS_RETRY_ERROR_TYPE_SERVER_ERROR,
+    /** Doesn't count against any budgets. This could be something like a 401 challenge in Http. */
+    AWS_RETRY_ERROR_TYPE_CLIENT_ERROR,
+};
+
+struct aws_retry_strategy_vtable {
+    void (*destroy)(struct aws_retry_strategy *retry_strategy);
+    int (*acquire_token)(
+        struct aws_retry_strategy *retry_strategy,
+        const struct aws_byte_cursor *partition_id,
+        aws_retry_strategy_on_retry_token_acquired_fn *on_acquired,
+        void *user_data,
+        uint64_t timeout_ms);
+    int (*schedule_retry)(
+        struct aws_retry_token *token,
+        enum aws_retry_error_type error_type,
+        aws_retry_strategy_on_retry_ready_fn *retry_ready,
+        void *user_data);
+    int (*record_success)(struct aws_retry_token *token);
+    void (*release_token)(struct aws_retry_token *token);
+};
+
+struct aws_retry_strategy {
+    struct aws_allocator *allocator;
+    struct aws_retry_strategy_vtable *vtable;
+    struct aws_atomic_var ref_count;
+    void *impl;
+};
+
+struct aws_retry_token {
+    struct aws_allocator *allocator;
+    struct aws_retry_strategy *retry_strategy;
+    void *impl;
+};
+
+/**
+ * Jitter mode for exponential backoff.
+ *
+ * For a great writeup on these options see:
+ * https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+ */
+enum aws_exponential_backoff_jitter_mode {
+    AWS_EXPONENTIAL_BACKOFF_JITTER_NONE,
+    AWS_EXPONENTIAL_BACKOFF_JITTER_FULL,
+    AWS_EXPONENTIAL_BACKOFF_JITTER_DECORRELATED,
+    /* Uses AWS_EXPONENTIAL_BACKOFF_JITTER_FULL */
+    AWS_EXPONENTIAL_BACKOFF_JITTER_DEFAULT,
+};
+
+struct aws_exponential_backoff_retry_config {
+    /** Event loop group to use for scheduling tasks. */
+    struct aws_event_loop_group *el_group;
+    /** Max retries to allow. This value must be greater than 0 */
+    size_t max_retries;
+    /** Scaling factor to add for the backoff. 25ms is usually a good default. */
+    uint32_t backoff_scale_factor_ms;
+    /** Jitter mode to use, see comments for aws_exponential_backoff_jitter_mode. */
+    enum aws_exponential_backoff_jitter_mode jitter_mode;
+    /** By default this will be set to use aws_device_random. If you want something else, set it here. */
+    uint64_t (*generate_random)(void);
+};
+
+/**
+ * Defaults to 10 retries, 25ms scaling factor, and full jitter.
+ */
+#define AWS_EXPONENTIAL_BACKOFF_DEFAULT_CONFIG(el_group)                                                               \
+    /* Don't do this with designated initializers, otherwise C++ users can't touch this. */                            \
+    { el_group, 10, 25, AWS_EXPONENTIAL_BACKOFF_JITTER_FULL, }
```

Comment:
debatable/trivial: could we just treat 0 as "defaults pls" for everything in the config?
but that strategy falls apart if 0 is a valid value for something

Reply:
we could.